Epic 8080 : WNPRC Purchasing Module - Initial Setup

labkey-bpatel opened the pull request:
#### Rationale
Initial implementation of WNPRC Purchasing workflow

#### Related Pull Requests
https://github.com/LabKey/ehrModules/pull/95
https://github.com/LabKey/platform/pull/174
https://github.com/LabKey/distributions/pull/116

#### Changes
- WNPRC Purchasing module classes 
- Remove unsupported properties
- Add QC Status values upon 'wnprc_purchasing' folder creation
- Add landing page
- Add react setup files

labkey-bpatel committed:
Remove unsupported properties

labkey-bpatel committed:
WNPRC Purchasing Module
- Add WNPRC_Purchasing folder type
- Add setup files for React dev
- Add landing page (wip)
- Add Request Entry page with wip message
- Other basic module setup

labkey-bpatel committed:
Add QCStatus values upon 'wnprc_purchasing' container creation.

labkey-bpatel committed:
Add a landing page with Purchase Requests view.
Add a button to 'Create Request'.
Configure button with an action url to route user to a react page (which is intended to be a Purchase Request data entry form).

labkey-bpatel committed:
Cleanup

labkey-jeckels commented on WNPRC_Purchasing/build.gradle:
```diff
@@ -0,0 +1,20 @@
+import org.labkey.gradle.util.BuildUtils
+
+dependencies {
+    implementation "net.sf.opencsv:opencsv:${opencsvVersion}"
```
I don't think any of these four dependencies are relevant at this point, correct?

labkey-bpatel replied:
correct, removed

labkey-jeckels commented on WNPRC_Purchasing/gradle.properties:
```diff
@@ -0,0 +1,3 @@
+javafpdfVersion=1.6.0
```
These version numbers aren't relevant either.

labkey-bpatel replied:
removed

labkey-jeckels commented on WNPRC_Purchasing/resources/views/purchasingLandingPage.html:
```diff
@@ -0,0 +1,25 @@
+<div id="<%=h('wnprc_purchasing-report')%>"></div>
+<script type="text/javascript">
+
+    Ext4.onReady(function()
```
I don't think we need the onReady() here. Let's avoid ExtJS dependencies when we can.

labkey-bpatel replied:
good point, removed

labkey-jeckels commented on WNPRC_Purchasing/resources/views/purchasingLandingPage.html:
```diff
@@ -0,0 +1,25 @@
+            schemaName: 'ehr_purchasing',
+            queryName: 'purchasingRequestsOverview',
+            buttonBar: {
+                includeStandardButtons: true,
```
Do we need to explicitly include all of the standard buttons?

labkey-bpatel replied:
Yes, otherwise it is listing the custom button first, and didn't see a property to set the location

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingModule.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public boolean hasScripts()
+    {
+        return true;
```
We don't have scripts yet for this module, correct? Let's return false until they're needed, and also return null from getSchemaVersion()

labkey-bpatel replied:
done

labkey-bpatel committed:
Code review changes:
- Metadata updates - omit fkColumnName where applicable, remove fkDisplayColumnName and set titleColumn instead
- Column updates - change to varchar, add FK and IDX to itemUnitId and itemStatusId
- Other minor cleanup

labkey-bpatel committed:
Add sample data and automated test.

labkey-bpatel committed:
Issue 41769: Fix WNPRC module.properties files

labkey-bpatel committed:
Add metadata for userAccountAssociations table.

labkey-bpatel committed:
Annotate test with @Category({CustomModules.class, EHR.class})

labkey-jeckels commented on WNPRC_Purchasing/test/src/org/labkey/test/tests/wnprc_purchasing/WNPRC_PurchasingTest.java:
```diff
@@ -0,0 +1,188 @@
+        _containerHelper.createProject(getProjectName(), FOLDER_TYPE);
+
+        log("Creating a purchasing admin user");
+        _adminUserId = _userHelper.createUser(ADMIN_USER).getUserId().toString();
```
Make the _adminUserId and _requesterUserId fields int instead of String, and use getUserId().intValue().

labkey-bpatel replied:
good point, done!

labkey-jeckels reviewed (approved):
One more suggestion for the newly added test coverage, but no additional review required.

labkey-bpatel committed:
Test review update.

labkey-sweta committed:
Removing the CUSTOM category

labkey-tchad commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingContainerListener.java:
```diff
@@ -0,0 +1,57 @@
+
+public class WNPRC_PurchasingContainerListener implements ContainerListener
+{
+    @Override
+    public void containerCreated(Container c, User user)
+    {
+        //populate qc state data upon WNPRC_Purchasing folder creation
+        WNPRC_PurchasingManager.get().addQCStatus(c, user);
+    }
```
I believe this will apply to all newly created containers, regardless of folder type. Also, it wouldn't apply to existing folders that have their type changed.

labkey-jeckels replied:
Good point Trey, I was thinking this was on a folder type but it's on the ContainerListener.

@labkey-bpatel I suggest introducing a FolderType implementation and doing this setup in its configureContainer() instead.

labkey-bpatel replied:
Thank you for catching that, Trey! I missed that it would add data regardless of folder type.
Now updated!

labkey-jeckels reviewed (changes requested):
See Trey's comments about the ContainerListener behavior.

labkey-bpatel committed:
Code review changes:
- Remove folderType.xml, instead create java class, and populate initial data.
- Add itemUnits and lineItemStatus data during folder creation.
- Update test to add WNPRC_Purchasing webpart, and remove uploading of itemUnits and lineItemStatus data.
- Add WNPRC_EHR test category.
- Remove react generated files pushed by mistake.

labkey-bpatel committed:
Code review changes:
-Remove adding of initial data from container listener
